src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/Authenticator.cs

```diff
@@ -90,26 +88,29 @@ public Authenticator(string authority, bool validateAuthority)
 
         public string SelfSignedJwtAudience { get; private set; }
 
+        public string PreferredCache { get; private set; }
+
+        public string[] Aliases { get; private set; }
+
         public Guid CorrelationId { get; set; }
 
         public async Task UpdateFromTemplateAsync(CallState callState)
         {
             if (!this.updatedFromTemplate)
             {
                 var authorityUri = new Uri(this.Authority);
-                string host = authorityUri.Authority;
                 string path = authorityUri.AbsolutePath.Substring(1);
                 string tenant = path.Substring(0, path.IndexOf("/", StringComparison.Ordinal));
-
-                AuthenticatorTemplate matchingTemplate = await AuthenticatorTemplateList.FindMatchingItemAsync(this.ValidateAuthority, host, tenant, callState).ConfigureAwait(false);
-
-                this.AuthorizationUri = matchingTemplate.AuthorizeEndpoint.Replace("{tenant}", tenant);
-                this.DeviceCodeUri = matchingTemplate.DeviceCodeEndpoint.Replace("{tenant}", tenant);
-                this.TokenUri = matchingTemplate.TokenEndpoint.Replace("{tenant}", tenant);
-                this.UserRealmUri = CanonicalizeUri(matchingTemplate.UserRealmEndpoint);
+                var entry = await InstanceDiscovery.GetMetadataEntry(authorityUri.Host, this.ValidateAuthority, callState);
+                this.AuthorizationUri = InstanceDiscovery.AuthorizeEndpointTemplate.Replace("{host}", entry.PreferredNetwork).Replace("{tenant}", tenant);
+                this.DeviceCodeUri = "https://{host}/{tenant}/oauth2/devicecode".Replace("{host}", entry.PreferredNetwork).Replace("{tenant}", tenant);
+                this.TokenUri = "https://{host}/{tenant}/oauth2/token".Replace("{host}", entry.PreferredNetwork).Replace("{tenant}", tenant);
+                this.UserRealmUri = CanonicalizeUri("https://{host}/common/UserRealm".Replace("{host}", entry.PreferredNetwork));
                 this.IsTenantless = (string.Compare(tenant, TenantlessTenantName, StringComparison.OrdinalIgnoreCase) == 0);
-                this.SelfSignedJwtAudience = matchingTemplate.Issuer.Replace("{tenant}", tenant);
+                this.SelfSignedJwtAudience = this.TokenUri;
```

Comment:
Is matchingTemplate.Issuer same as tokenUri? if not, then why the change?

Reply:
It was essentially coded with the same effect, according to the previous existing implementation.

The `matchingTemplate.Issuer` was the same as its own TokenEndpiont, according to [this line](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/796/files#diff-e4febd8f40f03e71bcae0f990f9690eaL52). And then we updated it with current tenant. All these were exactly the same as how we calculated the `TokenUri` in [this line](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/796/files#diff-b8caab14db87a8ac7e70708851e39df7L108).

Providing that the original implementation was correct (please help double check), the latest implementation is equivalent and hopefully more clear about the intention.

Reply:
got it, thanks